src/Polysemy/ConstraintAbsorber/MonadWriter.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE FlexibleInstances           #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving  #-}
+{-# LANGUAGE MultiParamTypeClasses       #-}
+{-# LANGUAGE ScopedTypeVariables         #-}
+{-# LANGUAGE UndecidableInstances        #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
+
+module Polysemy.ConstraintAbsorber.MonadWriter
+  (
+    absorbWriter
+  )
+where
+
+import           Polysemy
+import           Polysemy.ConstraintAbsorber
+import           Polysemy.Writer
+import qualified Control.Monad.Writer.Class as S
+
+------------------------------------------------------------------------------
+-- | absorb a @MonadWriter w@ constraint into @Member (Writer w) r => Sem r@
+absorbWriter :: forall w r a. (Monoid w
+                              ,Member (Writer w) r)
+  => (S.MonadWriter w (Sem r) => Sem r a) -> Sem r a
+absorbWriter =
+  let semTell = tell
+      semListen :: Member (Writer w) r => Sem r b -> Sem r (b, w) 
+      semListen = fmap (\(x,y) -> (y,x)) . listen @w
+      semPass ::  Member (Writer w) r => Sem r (b, w -> w) -> Sem r b
+      semPass x = do
+        (w, (a, f)) <- listen x
+        censor f (tell w >> pure a)
+  in absorbWithSem @(S.MonadWriter _) @Action 
+     (WriterDict semTell semListen semPass)
+     (Sub Dict)  
```

Comment:
Is it possible to factor this `Sub Dict` into `absorbWithSem` or does it break the constraint machinery somehow?

Reply:
We could, I think.  But I'm worried there will come a time when someone needs to do something else there, like a more complex entailment or something and not having access to that argument will be annoying.  But I don't know since my brain frizzes at ```Sub Dict`` ` in the first place.  Should I try to have two versions of ```absorbWithSem```, one simplified and one where you can access that argument just in case?

Reply:
Actually, right now I can't make that work anyway.  Once I check in the rest of these fixes and you take a look, we can decide if it's worth chasing.

Reply:
Sounds good. I'm not convinced it's possible --- `Dict`s are weird --- but it's a noble goal if tenable.

Reply:
"Dicts are weird" seems like a good first album name for a very very geeky punk band.